influxdb_store.go

```diff
@@ -0,0 +1,331 @@
+package appdash
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"reflect"
+	"time"
+
+	influxDBClient "github.com/influxdb/influxdb/client"
+	influxDBServer "github.com/influxdb/influxdb/cmd/influxd/run"
+	influxDBModels "github.com/influxdb/influxdb/models"
+)
+
+const (
+	dbName               string = "appdash" // InfluxDB db name.
+	spanMeasurementName  string = "spans"   // InfluxDB container name for trace spans.
+	defaultTracesPerPage int    = 10        // Default number of traces per page.
+)
+
+// Compile-time "implements" check.
+var _ interface {
+	Store
+	Queryer
+} = (*InfluxDBStore)(nil)
+
+// TODO: should be a constant.
+var zeroID = fmt.Sprintf("%016x", uint64(0))
+
+type InfluxDBStore struct {
+	con           *influxDBClient.Client // InfluxDB client connection.
+	server        *influxDBServer.Server // InfluxDB API server.
+	tracesPerPage int                    // Number of traces per page.
+}
+
+func (in *InfluxDBStore) Collect(id SpanID, anns ...Annotation) error {
+	// Current strategy is to remove existing span and save new one
+	// instead of updating the existing one.
+	// TODO: explore a more efficient alternative strategy.
+	if err := in.removeSpanIfExists(id); err != nil {
+		return err
+	}
+	// trace_id, span_id & parent_id are set as tags
+	// because InfluxDB tags are indexed & those values
+	// are uselater on queries.
+	tags := make(map[string]string, 3)
+	tags["trace_id"] = id.Trace.String()
+	tags["span_id"] = id.Span.String()
+	tags["parent_id"] = id.Parent.String()
+	// Saving annotations as InfluxDB measurement spans fields
+	// which are not indexed.
+	fields := make(map[string]interface{}, len(anns))
+	for _, ann := range anns {
+		fields[ann.Key] = string(ann.Value)
+	}
+	// InfluxDB point represents a single span.
+	pts := []influxDBClient.Point{
+		influxDBClient.Point{
+			Measurement: spanMeasurementName,
+			Tags:        tags,   // indexed metadata.
+			Fields:      fields, // non-indexed metadata.
+			Time:        time.Now(),
+			Precision:   "s",
```

Comment:
FYI we should default to `ms` precision not `s`. Most of our measurements will be in `ms` I think. If i'm wrong about how this relates to InfluxDB just let me know.